Mathlib/NumberTheory/ModularForms/EisensteinSeries/UniformConvergence.lean

```diff
@@ -0,0 +1,291 @@
+  · apply add_nonneg (pow_two_nonneg _) ?_
+    apply mul_nonneg
+    · linarith
+    · apply pow_two_nonneg
+  · positivity
+
+lemma auxbound1 (z : ℍ) {δ : ℝ} (ε : ℝ) (hδ : 1 ≤ δ ^ 2) : r z ≤ Complex.abs (δ * (z : ℂ) + ε) := by
+  rw [r, Complex.abs]
+  have H1 : (z : ℂ).im ≤
+    Real.sqrt ((δ * (z : ℂ).re + ε) * (δ * (z : ℂ).re + ε) + (δ * z : ℂ).im * (δ * z : ℂ).im) := by
+    have h1 : (δ * z : ℂ).im * (δ * z : ℂ).im = δ ^ 2 * (z : ℂ).im * (z : ℂ).im := by
+      simp only [mul_im, ofReal_re, coe_im, ofReal_im, coe_re, zero_mul, add_zero]
+      ring
+    rw [Real.le_sqrt', h1]
+    · nlinarith
+    · exact z.2
+  simp only [UpperHalfPlane.coe_im, UpperHalfPlane.coe_re, AbsoluteValue.coe_mk, MulHom.coe_mk,
+    min_le_iff] at *
+  left
+  simp only [one_le_sq_iff_one_le_abs, mul_im, ofReal_re, coe_im, ofReal_im, coe_re, zero_mul,
+    add_zero, normSq_apply, add_re, mul_re, sub_zero, add_im] at *
+  exact H1
+
+lemma auxbound2 (z : ℍ) (δ : ℝ) {ε : ℝ} (hε : 1 ≤ ε ^ 2) : r z ≤ Complex.abs (δ * (z : ℂ) + ε) := by
+  rw [r, Complex.abs, min_le_iff]
+  have H1 : Real.sqrt (r1 z) ≤ Real.sqrt ((δ * (z : ℂ).re + ε) * (δ * (z : ℂ).re + ε) +
+      δ * (z : ℂ).im * (δ * (z : ℂ).im)) := by
+    rw [r1, Real.sqrt_le_sqrt_iff, ← pow_two, ← pow_two]
+    apply r1_aux_bound z δ hε
+    nlinarith
+  right
+  simp only [ne_eq, coe_re, coe_im, normSq_apply, AbsoluteValue.coe_mk, MulHom.coe_mk, add_re,
+    mul_re, ofReal_re, ofReal_im, zero_mul, sub_zero, add_im, mul_im, add_zero] at *
+  exact H1
+
+lemma ne_zero_if_max {x : Fin 2 → ℤ} (hx : x ≠ 0)
+    (h : max (x 0).natAbs (x 1).natAbs = (x 0).natAbs) : x 0 ≠ 0 := by
+  contrapose! hx
+  ext x
+  fin_cases x <;> aesop
+
+/-This proof is faster than the one above, but longer. -/
+lemma ne_zero_if_max' {x : Fin 2 → ℤ} (hx : x ≠ 0)
+    (h : (max (x 0).natAbs (x 1).natAbs) = (x 1).natAbs) : (x 1) ≠ 0 := by
+  apply ne_zero_if_max (x := ![x 1, x 0]) ?_ (by simpa only [Fin.isValue, Matrix.cons_val_zero,
+    Matrix.cons_val_one, Matrix.head_cons, max_eq_left_iff, max_eq_right_iff] using h)
+  simp only [ne_eq, Matrix.cons_eq_zero_iff, Matrix.zero_empty, and_true, not_and]
+  intro h1 h0
+  rw [Function.ne_iff, Fin.exists_fin_two, h1, h0] at hx
+  simp only [Pi.zero_apply, ne_eq, not_true_eq_false, or_self] at *
+
+lemma div_max_sq_ge_one (x : Fin 2 → ℤ) (hx : x ≠ 0) :
+    (1 : ℝ) ≤ (x 0 / (max (x 0).natAbs (x 1).natAbs)) ^ 2 ∨
+      (1 : ℝ) ≤ (x 1 / (max (x 0).natAbs (x 1).natAbs)) ^ 2 := by
+  refine (max_choice (x 0).natAbs (x 1).natAbs).imp (fun H1 => ?_) (fun H2 => ?_)
+  · simp only [Fin.isValue, H1, Int.cast_natAbs, Int.cast_abs, div_pow, _root_.sq_abs, ne_eq,
+    OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero, ne_zero_if_max hx H1,
+    div_self, le_refl]
+  · simp only [Fin.isValue, H2, Int.cast_natAbs, Int.cast_abs, div_pow, _root_.sq_abs, ne_eq,
+    OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero,
+    ne_zero_if_max' hx H2, div_self, le_refl]
```

Comment:
alternative suggestion (not much difference to the original really):
```suggestion
lemma div_max_sq_ge_one (x : Fin 2 → ℤ) (hx : x ≠ 0) :
    (1 : ℝ) ≤ (x 0 / (max (x 0).natAbs (x 1).natAbs)) ^ 2 ∨
      (1 : ℝ) ≤ (x 1 / (max (x 0).natAbs (x 1).natAbs)) ^ 2 := by
  rcases max_choice (x 0).natAbs (x 1).natAbs with H1 | H2
  · simp only [H1, Int.cast_natAbs, Int.cast_abs, div_pow, _root_.sq_abs, ne_eq,
      OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero,
      left_ne_zero_of_max_eq hx H1, div_self, le_refl, true_or]
  · simp only [H2, Int.cast_natAbs, Int.cast_abs, div_pow, _root_.sq_abs, ne_eq,
      OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero,
      right_ne_zero_of_max_eq hx H2, div_self, le_refl, or_true]
```

Reply:
The original one is marginally faster, so I'll keep it (modulo the name changes needed).